searchlib/src/tests/diskindex/pagedict4/pagedict4_hugeword_cornercase_test.cpp

```diff
@@ -0,0 +1,205 @@
+{
+    PostingListCounts::Segment lastseg = counts._segments.back();
+    PostingListCounts::Segment fillseg;
+    fillseg._bitLength = 4000000;
+    fillseg._numDocs = minChunkDocs;
+    fillseg._lastDoc = minChunkDocs;
+    counts._bitLength += fillseg._bitLength;
+    counts._numDocs += fillseg._numDocs;
+    counts._segments.back() = fillseg;
+    uint32_t lastDoc = counts._segments.size() * fillseg._numDocs;
+    counts._segments.back()._lastDoc = lastDoc;
+    counts._segments.push_back(lastseg);
+    counts._segments.back()._lastDoc = lastDoc + lastseg._numDocs;
+}
+
+PostingListCounts makeBaseCounts()
+{
+    PostingListCounts counts;
+    PostingListCounts::Segment lastseg;
+    lastseg._bitLength = 100;
+    lastseg._numDocs = 10;
+    lastseg._lastDoc = 10;
+    counts._bitLength = lastseg._bitLength;
+    counts._numDocs = lastseg._numDocs;
+    counts._segments.push_back(lastseg);
+    addSegment(counts);
+    return counts;
+}
+
+PostingListCounts makeSegmentedCounts(uint32_t segments)
+{
+    PostingListCounts counts = makeBaseCounts();
+    while (counts._segments.size() < segments) {
+        addSegment(counts);
+    }
+    return counts;
+}
+
+uint32_t
+calcSegments(uint32_t maxLen)
+{
+    BitBuffer bb;
+    PostingListCounts counts = makeBaseCounts();
+    uint32_t len = bb.getSize(counts);
+    unsigned int i = 0;
+    while (len <= maxLen) {
+        addSegment(counts);
+        ++i;
+        len = bb.getSize(counts);
+    }
+    return counts._segments.size() - 1;
+}
+
+PostingListCounts makeCounts(uint32_t wantLen)
+{
+    BitBuffer bb;
+    uint32_t segments = calcSegments(wantLen);
+    PostingListCounts counts = makeSegmentedCounts(segments);
+    PostingListCounts counts2 = makeSegmentedCounts(segments - 1);
+    uint32_t len = bb.getSize(counts);
+    uint32_t len2 = bb.getSize(counts2);
+    for (unsigned int i = 1; i + 2 < counts._segments.size(); ++i) {
+        counts._bitLength += counts._segments[0]._bitLength;
+        counts._segments[i]._bitLength += counts._segments[0]._bitLength;
+        counts2._bitLength += counts2._segments[0]._bitLength;
+        counts2._segments[i]._bitLength += counts2._segments[0]._bitLength;
+        len = bb.getSize(counts);
+        len2 = bb.getSize(counts2);
+        if (len == wantLen) {
+            return counts;
+        }
+        if (len2 == wantLen) {
+            return counts2;
+        }
+    }
+    LOG(info, "Could not calculate counts with wanted compressed length");
+    abort();
+}
+
+using StartOffset = search::bitcompression::PageDict4StartOffset;
+using Writer = search::diskindex::test::PageDict4MemWriter;
+using SeqReader = search::diskindex::test::PageDict4MemSeqReader;
+using RandReader = search::diskindex::test::PageDict4MemRandReader;
+
+void testWords()
+{
+    using EC = search::bitcompression::PostingListCountFileEncodeContext;
+    using DC = search::bitcompression::PostingListCountFileDecodeContext;
+
+    EC pe;
```

Comment:
Also consider separating setup of data with Writer, construction of Reader and verification using Reader. This might solve much of the naming problems as well.

Reply:
Seems like this has been addressed in later commit where ownership of encode and decode
contexts has been moved to shared test classes.